### DISCUSSION TITLE
Multiple pages share state?

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#1756] [Q&A]
### Question

Hi, I have a question.  I'm testing the nicegui examples on my system.  I run the [Run Javascript](https://github.com/zauberzeug/nicegui) example verbatim and it opens a new tab in chrome at http://127.0.0.1:8000.  Click Fire and Forget, alert box pops up.  I dismiss the alert.  So far so good.

Next I try another browser.  I copy and paste the url http://127.0.0.1:8000 to Firefox.  It opens the same page, no problem. I click Fire and Forget, Firefix opens an alert box.  All as expected.  So what's wrong?

# Issue 
The issue is this: I switch back to chrome to the app page that was left open.  It too now shows an open alert box, without me touching anything.  Clicking the button once in Firefox opened 2 alert boxes: one in chrome page and one in firefox page.  Same thing if I open the page in Vivaldi.

It seems the nicegui app doesn't distinguish http client sessions?  Any open page to the nicegui URL controls the same instance.  Actions taken on one page are propagated to all open pages via the nicegui server.

# Questions
Is this expected behavior?  Are there ways to prevent this so each http client has a different instance of the page with its own state?  Can you run multiple clients from one nicegui server?  I didn't see any args to `run ()` that change this behavior.

Thanks for your help.
Edward

**Comment by falkoschindler** [comment]
This is expected behavior: All UI elements simply placed in the global scope are shared among all clients connected to "/". We call it the auto-index page.
To create private pages that are generated for each client individually, you should use functions with the `@page` decorator.
See https://nicegui.io/documentation#auto-index_page and https://nicegui.io/documentation#page for more information.

**Reply by ed2050** [reply]
Thanks Falko.  I guess I wasn't expecting updates from auto-index page to be pushed out to multiple clients.  Especially when the update is triggered by one client.  It's great for consistency though.

The thing I'm still not clear on is what triggers creation of a new page.  Is it one page per HTTP connection?  So if client hits reload, they get a brand new instance of the page?  Or is it one page per user?  The docs aren't clear to me on this point.

# Docs
[The auto-index docs](https://nicegui.io/documentation#auto-index_page) seem to indicate that every page reload creates a new page:
```the displayed ID on the private page changes when the browser reloads the page.```

But then how do you deal with interrupted connections?  How do you handle user logins, where connections may be interrupted or pages closed, but the same user opens the same page again?  Is the whole page torn down and rebuilt each time?  Does that mean any persistent state has to be explicitly saved and reloaded by the app?

Under [Storage](https://nicegui.io/documentation#storage), the docs suggest that nicegui can id users across HTTP connections with browser cookies:

```app.storage.user:
Stored server-side, each dictionary is associated with a unique identifier 
held in a browser session cookie. Unique to each user, this storage is accessible 
across all their browser tabs. app.storage.browser['id'] is used to identify the user.
```
However it's not clear whether this tracking is handled at the server level to determine when to create new pages, or if tracking has to be done manually in the app code by examining app.storage.user.

Are there any examples that fit all this together?  Multiple users with multiple pages and persistent state.
Thanks again.

**Reply by rodja** [reply]
> But then how do you deal with interrupted connections?

The auto-index page stores the full state on the server. When a browser wants to load it (even when reconnecting) the full state is send to it upon initalizing.

> Are there any examples that fit all this together? Multiple users with multiple pages and persistent state.

Sure. Have a look at our auth-example: https://github.com/zauberzeug/nicegui/blob/main/examples/authentication/main.py
